### PR TITLE
added hashcode

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR introduces SIValue_HashCode for SIValue.

### CODE REVIEW

File: src/value.c [hunk 1]
```diff
@@ -469,6 +470,85 @@ int SIValue_Compare(const SIValue a, const SIValue b, int *disjointOrNull) {
...
+		break;
+	case T_STRING:
+		XXH64_update(&state, &v.stringval, strlen(v.stringval));
+		break;
+	case T_BOOL:
+	case T_INT64:
+		XXH64_update(&state, &v.longval, sizeof(v.longval));
+		break;
+	case T_DOUBLE: {
+		int64_t casted = (int64_t) v.doubleval;
```
Comment (swilly22) [inline src/value.c]: Explain why

File: tests/unit/test_hash_code.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	ASSERT_NE(origHashCode, otherHashCode);
+
+	siBoolOther = SI_BoolVal(false);
+	otherHashCode = SIValue_HashCode(siBoolOther);
+	ASSERT_EQ(origHashCode, otherHashCode);
+}
+
+TEST_F(HashCodeTest, TestHashLong) {
+	// INT32 and INT64 tests take too long.
+	for(int64_t i = INT16_MIN; i < INT16_MAX; i++) {
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply tests/unit/test_hash_code.cpp]: done

File: tests/unit/test_hash_code.cpp [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	siDouble = SI_DoubleVal(3.14159265);
+	origHashCode = SIValue_HashCode(siDouble);
+	ASSERT_NE(origHashCode, otherHashCode);
+
+	siDoubleOther = SI_DoubleVal(3.14159265);
+	otherHashCode = SIValue_HashCode(siDoubleOther);
+	ASSERT_EQ(origHashCode, otherHashCode);
+}
+
+
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply tests/unit/test_hash_code.cpp]: fixed

File: tests/unit/test_hash_code.cpp [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+	uint64_t otherHashCode = SIValue_HashCode(arrOther);
+	ASSERT_EQ(origHashCode, otherHashCode);
+
+	SIArray_Append(&arr, SI_LongVal(1));
+	origHashCode = SIValue_HashCode(arr);
+	ASSERT_NE(origHashCode, otherHashCode);
+
+	SIArray_Append(&arrOther, SI_LongVal(1));
+	otherHashCode = SIValue_HashCode(arrOther);
+	ASSERT_EQ(origHashCode, otherHashCode);
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply tests/unit/test_hash_code.cpp]: done

File: tests/unit/test_hash_code.cpp [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+/* Test for difference in hash code for the same binary representation
+ * for different types.*/
+TEST_F(HashCodeTest, TestHashLongAndBool) {
+	SIValue siInt64 = SI_LongVal(1);
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply tests/unit/test_hash_code.cpp]: done

File: src/value.h [hunk 6]
```diff
@@ -156,6 +156,9 @@ SIValue SIValue_Divide(const SIValue a, const SIValue b);
  * If the the values are not of the same type, the macro DISJOINT is returned in disjointOrNull value. */
 int SIValue_Compare(const SIValue a, const SIValue b, int *disjointOrNull);
 
+/* Returns a hash code for a given SIValue */
+uint64_t SIValue_HashCode(const SIValue v);
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/value.h]: done

File: src/datatypes/array.h [hunk 7]
```diff
@@ -51,9 +51,11 @@ SIValue SIArray_Clone(SIValue siarray);
   */
 void SIArray_ToString(SIValue list, char **buf, size_t *bufferLen, size_t *bytesWritten);
 
+uint64_t SIArray_HashCode(SIValue siarray);
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/datatypes/array.h]: done

File: src/datatypes/array.c [hunk 8]
```diff
@@ -59,6 +59,18 @@ void SIArray_ToString(SIValue list, char **buf, size_t *bufferLen, size_t *bytes
 	*bytesWritten += snprintf(*buf + *bytesWritten, *bufferLen, "]");
 }
 
+/* This method referenced by Java ArrayList.hashCode() method, which takes
+ * into account the hasing of nested values.*/
+uint64_t SIArray_HashCode(SIValue siarray) {
+	uint64_t hashCode = 1;
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/datatypes/array.c]: done

File: src/value.c [hunk 9]
```diff
@@ -469,6 +470,85 @@ int SIValue_Compare(const SIValue a, const SIValue b, int *disjointOrNull) {
...
+/* Hashes the id and properties of the node*/
+uint64_t SINode_HashCode(const SIValue v) {
+	XXH_errorcode res;
+	XXH64_state_t state;
+	res = XXH64_reset(&state, 0);
+	assert(res != XXH_ERROR);
+
+	Node *n = (Node *)v.ptrval;
+	int id = ENTITY_GET_ID(n);
+	XXH64_update(&state, &id, sizeof(id));
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/value.c]: done

File: src/value.c [hunk 10]
```diff
@@ -469,6 +470,85 @@ int SIValue_Compare(const SIValue a, const SIValue b, int *disjointOrNull) {
...
+	XXH64_state_t state;
+	res = XXH64_reset(&state, 0);
+	assert(res != XXH_ERROR);
+
+	Edge *e = (Edge *)v.ptrval;
+	int id = ENTITY_GET_ID(e);
+	XXH64_update(&state, &id, sizeof(id));
+	XXH64_update(&state, &(v.type), sizeof(v.type));
+
+	uint64_t hashCode =  XXH64_digest(&state);
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/value.c]: done

File: src/value.c [hunk 11]
```diff
@@ -469,6 +470,85 @@ int SIValue_Compare(const SIValue a, const SIValue b, int *disjointOrNull) {
...
+	XXH_errorcode res;
+	XXH64_state_t state;
+	res = XXH64_reset(&state, 0);
+	assert(res != XXH_ERROR);
+
+	Node *n = (Node *)v.ptrval;
+	int id = ENTITY_GET_ID(n);
+	XXH64_update(&state, &id, sizeof(id));
+	XXH64_update(&state, &(v.type), sizeof(v.type));
+	uint64_t hashCode =  XXH64_digest(&state);
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/value.c]: done

File: src/value.c [hunk 12]
```diff
@@ -469,6 +470,85 @@ int SIValue_Compare(const SIValue a, const SIValue b, int *disjointOrNull) {
...
+
+/* This method hashes primitive types in place. Compound types have their own
+ * hashing method */
+uint64_t SIValue_HashCode(const SIValue v) {
+	XXH_errorcode res;
+	XXH64_state_t state;
+	res = XXH64_reset(&state, 0);
+	assert(res != XXH_ERROR);
+
+	SIType t = SI_NUMERIC & v.type ? SI_NUMERIC : v.type;
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/value.c]: done

File: src/value.c (hunk 1 above)
Comment (DvirDukhan) [reply src/value.c]: done

**Pull request merged**